File: common/src/main/java/com/cainsgl/common/entity/user/UserEntity.java
package com.cainsgl.common.entity.user;

import com.baomidou.mybatisplus.annotation.*;
import com.baomidou.mybatisplus.extension.handlers.FastjsonTypeHandler;
import com.cainsgl.common.handler.StringListTypeHandler;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.apache.ibatis.type.ArrayTypeHandler;
import org.apache.ibatis.type.JdbcType;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@AllArgsConstructor
@TableName("users")
public class UserEntity {

    @TableId(type = IdType.ASSIGN_ID)
    private Long id;

    @TableField("username")
    private String username;

    @TableField("email")
    private String email;

    @TableField("password_hash")
    private String passwordHash;

    @TableField("nickname")
    private String nickname;

    @TableField("avatar_url")
    private String avatarUrl;

    @TableField("bio")
    private String bio;

    @TableField("level")
    private Integer level;

    @TableField("experience")
    private Integer experience;

    @TableField(value = "roles", typeHandler = StringListTypeHandler.class)
    private List<String> roles;

    @TableField(value = "permissions", typeHandler = StringListTypeHandler.class)
    private List<String> permissions;

    @TableField("status")
    private String status;

    @TableField("email_verified")
    private Boolean emailVerified;

    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private OffsetDateTime createdAt;

    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    private OffsetDateTime updatedAt;
    //到下一级的总经验值
    @TableField(exist = false)
    private Integer nextLevelTotalExp;
    //从现在到下一级需要的经验值
    @TableField(exist = false)
    private Integer expToNextLevel;
    //计算对应的成员变量
    public void calculateLevelInfo() {
        if (this.level != null && this.experience != null) {
            this.nextLevelTotalExp = (int) (Math.pow(2, this.level + 1) + this.level);
            this.expToNextLevel = Math.max(0, this.nextLevelTotalExp - this.experience);
        }
    }
    //去除敏感字段
    public void removeSensitiveFields() {
        this.passwordHash = null;
         this.emailVerified = null;
         this.status = null;
    }
}

File: common/src/main/java/com/cainsgl/common/entity/user/UserEntity.kt
package com.cainsgl.common.entity.user

import com.alibaba.fastjson2.annotation.JSONField
import com.baomidou.mybatisplus.annotation.*
import com.cainsgl.common.handler.StringListTypeHandler
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.fasterxml.jackson.databind.ser.std.ToStringSerializer
import java.time.LocalDateTime
import kotlin.math.max
import kotlin.math.pow

@TableName("\"user\"")
data class UserEntity(
    @TableId(type = IdType.ASSIGN_ID)
    @field:JsonSerialize(using = ToStringSerializer::class)
    var id: Long? = null,

    @TableField("username")
    var username: String? = null,

    @TableField("email")
    var email: String? = null,

    @field:JSONField(serialize = false)
    @TableField("password_hash")
    var passwordHash: String? = null,

    @TableField("nickname")
    var nickname: String? = null,

    @TableField("avatar_url")
    var avatarUrl: String? = null,

    @TableField("bio")
    var bio: String? = null,

    @TableField("level")
    var level: Int? = null,

    @TableField("gender")
    var gender: String? = null,

    @TableField("experience")
    var experience: Int? = null,

    @TableField(value = "roles", typeHandler = StringListTypeHandler::class)
    var roles: List<String>?=null,

    @TableField(value = "permissions", typeHandler = StringListTypeHandler::class)
    var permissions: List<String>?=null,

    @field:JSONField(serialize = false)
    @TableField("status")
    var status: String? = null,
    @TableField("email_verified")
    var emailVerified: Boolean? = null,
    @TableField("phone")
    var phone: String? = null,
    @TableField("used_memory")
    var usedMemory: Int? = null,
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    var createdAt: LocalDateTime? = null,
    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    var updatedAt: LocalDateTime? = null,
    @TableField(value = "extra", select = false)
    var extra: String? = null
)
{
    companion object{
        val BASIC_COL= listOf("id","nickname","avatar_url","level","gender")
        val DEFAULT_ROLE=listOf("user")
        val DEFAULT_PERMISSIONS=emptyList<String>()
    }

    //到下一级的总经验值
    @TableField(exist = false)
    var nextLevelTotalExp: Int? = null
    //从现在到下一级需要的经验值
    @TableField(exist = false)
    var expToNextLevel: Int? = null


    //计算对应的成员变量
    fun calculateLevelInfo(): UserEntity
    {
        if(this.level != null&&this.experience!=null)
        {
            this.nextLevelTotalExp = (2.0.pow(this.level!! + 1) + this.level!!).toInt()
            this.expToNextLevel = max(0, this.nextLevelTotalExp!! - this.experience!!)
        }
        return this
    }

    //去除系统级别的字段，比如密码
    fun sanitizeSystemSensitiveData(): UserEntity
    {
        this.passwordHash = null
        return this
    }

    fun isActive(): Boolean
    {
        return "active" == this.status
    }

    data class Extra(var bannedTime: LocalDateTime? = null)
}
